### PR TITLE
[libra-dev] use metadata script for crafting transactions

### PR DESCRIPTION
**Pull request by phoenix-antigravity**

### CODE REVIEW

File: client/libra-dev/src/transaction.rs [hunk 1]
```diff
@@ -588,11 +638,12 @@ mod test {
         let expiration_time_secs = 5;
         let signature = Ed25519Signature::try_from(&[1u8; Ed25519Signature::LENGTH][..]).unwrap();
 
-        let program = encode_transfer_script(
+        let program = encode_transfer_with_metadata_script(
```
Comment (christinacelee) [inline client/libra-dev/src/transaction.rs:641]: is it possible to add tests for this with actual metadata script?

File: client/libra-dev/src/transaction.rs [hunk 2]
```diff
@@ -343,6 +361,32 @@ pub unsafe extern "C" fn libra_LibraSignedTransaction_from(
...
+                            metadata_bytes: vec![].as_ptr(),
+                            metadata_len: 0,
+                        },
+                    });
+                } else {
+                    update_last_error("Fail to decode transaction payload".to_string());
+                    return LibraStatus::InternalError;
+                }
+            }
+            "peer_to_peer_with_metadata_transaction" => {
```
Comment (sblackshear) [inline client/libra-dev/src/transaction.rs:373]: Should we remove support for `peer_to_peer_transaction` (maybe just kill that script altogether) and always use `with_metadata`? The two are identical except that `peer_to_peer_transaction` fills in an empty metadata for you.

Comment (sblackshear) [reply client/libra-dev/src/transaction.rs:373]: Sweet. I will remove the script after you land this.

Comment (phoenix-antigravity) [reply client/libra-dev/src/transaction.rs:373]: it requires a bit more than this PR. For example, CLI is not using this FFI layer. I can migrate other call sites in separate PR

Comment (thefallentree) [reply client/libra-dev/src/transaction.rs:373]: @sblackshear  strongly agree!

File: client/libra-dev/include/data.h [hunk 3]
```diff
@@ -122,10 +124,12 @@ enum LibraStatus libra_LibraAccountResource_from(const uint8_t *buf, size_t len,
...
+ * @param[in] metadata_bytes is the metadata bytes for given transaction.
+ * @param[in] metadata_len is the length of metadata_bytes array.
  * @param[out] ptr_buf is the pointer that will be filled with the memory address of the transaction allocated in rust. User takes ownership of pointer returned by *buf, which needs to be freed using libra_free_bytes_buffer
  * @param[out] ptr_len is the length of the signed transaction memory buffer.
 */
-enum LibraStatus libra_SignedTransactionBytes_from(const uint8_t sender_private_key[LIBRA_PRIVKEY_SIZE], const uint8_t receiver[LIBRA_PUBKEY_SIZE], uint64_t sequence, uint64_t num_coins, uint64_t max_gas_amount, uint64_t gas_unit_price, uint64_t expiration_time_secs, uint8_t **ptr_buf, size_t *ptr_len);
+enum LibraStatus libra_SignedTransactionBytes_from(const uint8_t sender_private_key[LIBRA_PRIVKEY_SIZE], const uint8_t receiver[LIBRA_PUBKEY_SIZE], uint64_t sequence, uint64_t num_coins, uint64_t max_gas_amount, uint64_t gas_unit_price, uint64_t expiration_time_secs, const uint8_t metadata_bytes, uint8_t metadata_len, uint8_t **ptr_buf, size_t *ptr_len);
```
Comment (thefallentree) [inline client/libra-dev/include/data.h:132]: the header should be const uint8_t* metadata_bytes

File: client/libra-dev/include/data.h [hunk 4]
```diff
@@ -156,10 +160,12 @@ enum LibraStatus libra_LibraSignedTransaction_from(const uint8_t *buf, size_t le
...
  * @param[in] expiration_time_secs is the time this TX remain valid, the format is unix timestamp.
+ * @param[in] metadata_bytes is the metadata bytes for given transaction.
+ * @param[in] metadata_len is the length of metadata_bytes array.
  * @param[out] buf is the pointer that will be filled with the memory address of the transaction allocated in rust. User takes ownership of pointer returned by *buf, which needs to be freed using libra_free_bytes_buffer
  * @param[out] len is the length of the raw transaction memory buffer.
 */
-enum LibraStatus libra_RawTransactionBytes_from(const uint8_t sender[LIBRA_ADDRESS_SIZE], const uint8_t receiver[LIBRA_PUBKEY_SIZE], uint64_t sequence, uint64_t num_coins, uint64_t max_gas_amount, uint64_t gas_unit_price, uint64_t expiration_time_secs, uint8_t **buf, size_t *len);
+enum LibraStatus libra_RawTransactionBytes_from(const uint8_t sender[LIBRA_ADDRESS_SIZE], const uint8_t receiver[LIBRA_PUBKEY_SIZE], uint64_t sequence, uint64_t num_coins, uint64_t max_gas_amount, uint64_t gas_unit_price, uint64_t expiration_time_secs, const uint8_t metadata_bytes, uint8_t metadata_len, uint8_t **buf, size_t *len);
```
Comment (thefallentree) [inline client/libra-dev/include/data.h:168]: same here: the header should be const uint8_t* metadata_bytes

unfortunately we don't' have test coverage here.

**Comment by thefallentree** [comment]
Thanks!

**Pull request closed without merging**